query_executor.go

```diff
@@ -50,51 +70,148 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 			continue
 		}
 
-		iter = q.attemptQuery(qry, conn)
-		// Update host
-		hostResponse.Mark(iter.err)
+		// store the result
+		q.specWG.Add(1)
+		go q.runAndStore(qry, conn, selectedHost, results, nil)
+
+		// fetch from the channel whatever is there
+		res = <-results
 
-		if rt == nil {
-			iter.host = host
-			break
+		if res.retryType == RetryNextHost {
+			continue
 		}
 
-		switch rt.GetRetryType(iter.err) {
-		case Retry:
-			for rt.Attempt(qry) {
-				iter = q.attemptQuery(qry, conn)
-				hostResponse.Mark(iter.err)
-				if iter.err == nil {
-					iter.host = host
-					return iter, nil
-				}
-				if rt.GetRetryType(iter.err) != Retry {
-					break
-				}
-			}
-		case Rethrow:
-			return nil, iter.err
-		case Ignore:
-			return iter, nil
-		case RetryNextHost:
-		default:
+		return res.iter, res.err
+
+	}
+
+	if res.iter == nil {
+		return nil, ErrNoConnections
+	}
+
+	return res.iter, nil
+}
+
+func (q *queryExecutor) executeSpeculativeQuery(qry ExecutableQuery) (*Iter, error) {
+	hostIter := q.policy.Pick(qry)
+	sp := qry.speculativeExecutionPolicy()
+
+	results := make(chan queryResponse)
+
+	stopExecutions := make(chan struct{})
+	defer close(stopExecutions)
+
+	var (
+		res             queryResponse
+		specExecCounter int
+	)
+	for selectedHost := hostIter(); selectedHost != nil && specExecCounter < sp.Executions(); selectedHost = hostIter() {
+		host := selectedHost.Info()
+		if host == nil || !host.IsUp() {
+			continue
 		}
 
-		// Exit for loop if the query was successful
-		if iter.err == nil {
-			iter.host = host
-			return iter, nil
+		pool, ok := q.pool.getPool(host)
+		if !ok {
+			continue
+		}
+
+		conn := pool.Pick()
+		if conn == nil {
+			continue
+		}
+
+		// if it's not the first attempt, pause
+		if specExecCounter > 0 {
+			<-time.After(sp.Delay())
```

Comment:
reuse a timer instead of using time.After. This should also have a way to cancel, for now just doing a select on the ctx.Done() should be enough. Though that would need exposing from the query

Reply:
Will do.